### DISCUSSION TITLE
Remote logging to Azure blob storage

### DISCUSSION THREAD

**Discussion by HakarK** [airflow-helm/charts#749] [Questions & Answers] [answered]
I am getting this response when I try to read the logs
![image](https://github.com/airflow-helm/charts/assets/60604135/c187ef6f-38b4-442c-9baf-46855f930e6e)

I have followed this guide for setup
![image](https://github.com/airflow-helm/charts/assets/60604135/3c8712a0-3785-43a7-bc3d-f12ef2925bd9)

I have created an Airflow connection called my_wasb which works (or atleast I am getting "Successfully connected to Azure Blob Storage." when i test the connection)

And i have a storage account called _airflowstvg_ and a container called _wasb-airflowlogs_.

The config section in values file looks like this
![image](https://github.com/airflow-helm/charts/assets/60604135/4321afae-7600-4568-9fb0-52b2d231cc1a)

Do I need to follow these steps also?
![image](https://github.com/airflow-helm/charts/assets/60604135/269fd47d-911f-4c49-9056-10e05d5d2e01)

**Comment by thesuperzapper** [comment]
@HakarK I think something must have changed in the Azure provider.

Before we try using  an `airflow_local_settings.py` ([which the chart provides values for](https://github.com/airflow-helm/charts/blob/main/charts/airflow/docs/faq/configuration/airflow-configs.md#airflow_local_settingspy)), can you try formatting `AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER` like `wasb://<container_name>@<storage_account>.blob.core.windows.net` to see if that works?

Because right now, it looks like Airflow is not even trying to use a remote provider (probably because of the missing `wasb://` scheme).

---

If that fixes it, we should make an update to both the chart's docs and the upstream airflow provider docs:
- https://github.com/airflow-helm/charts/blob/main/charts/airflow/docs/faq/monitoring/log-persistence.md
- https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-microsoft-azure/logging/index.rst

**Reply by HakarK** [reply]
I still get the same response as above when trying to follow the scheme wasb://<container_name>@<storage_account>.blob.core.windows.net
![image](https://github.com/airflow-helm/charts/assets/60604135/597a55a3-86d2-403b-84ad-49e1955b5e09)

My storage account is named **airflowstvg** and container name **wasb-airflowlogs**:
![image](https://github.com/airflow-helm/charts/assets/60604135/2fb0eca9-e6fe-4457-b2ce-bb321ba8deed)

**Comment by thesuperzapper** [comment, marked as answer]
@HakarK after reading the airflow implementation, I can confirm that Azure Blob Store logging is really messed up, as [explained here](https://github.com/apache/airflow/blob/2.6.1/airflow/providers/microsoft/azure/log/wasb_task_handler.py#L135-L139).

Effectively, the problems are:

1. the Azure Blob Store container name has been [hard-coded to `airflow-logs`](https://github.com/apache/airflow/blob/2.6.1/airflow/config_templates/airflow_local_settings.py#L254)
2. the `REMOTE_BASE_LOG_FOLDER` [must start with `wasb` for airflow to know you are using Azure Blob Store](https://github.com/apache/airflow/blob/2.6.1/airflow/config_templates/airflow_local_settings.py#L247) (which causes blob names to always have a prefix that starts with `wasb/`)

However, we can work around these issues to get logs stored in the Azure Blob Store container named `MY_CONTAINER`, with the prefix `CUSTOM_PREFIX/`.

## SOLUTION: Airflow 2.6

__If you are using airflow 2.6, you can set:__

 - `AIRFLOW__LOGGING__REMOTE_LOGGING` to `True`
 - `AIRFLOW__LOGGING__REMOTE_LOG_CONN_ID` to `"my_wasb"` (which must have "login" set to your storage account name)
 - `AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER` to `"wasb"` (which is only to tell airflow we are using Azure Blob Store)
 - `AIRFLOW__LOGGING__REMOTE_TASK_HANDLER_KWARGS` to `{"wasb_container": "MY_CONTAINER", "wasb_log_folder": "CUSTOM_PREFIX"}`

## SOLUTION: Airflow 2.0 - 2.5

__If you are using airflow 2.0 - 2.5, you can set:__

 - `AIRFLOW__LOGGING__REMOTE_LOGGING` to `True`
 - `AIRFLOW__LOGGING__REMOTE_LOG_CONN_ID` to `"my_wasb"` (which must have "login" set to your storage account name)
 - `AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER` to `"wasb"` (which is only to tell airflow we are using Azure Blob Store)
 - `AIRFLOW__LOGGING__LOGGING_CONFIG_CLASS` to `"airflow_local_settings.LOGGING_CONFIG"`

__Then, you must create the following `airflow_local_settings.py` using the `airflow.localSettings.stringOverride` value:__

```yaml
airflow:
  localSettings:
    stringOverride: |
      from copy import deepcopy
      from airflow.config_templates.airflow_local_settings import DEFAULT_LOGGING_CONFIG

      LOGGING_CONFIG = deepcopy(DEFAULT_LOGGING_CONFIG)

      LOGGING_CONFIG["handlers"]["task"]["wasb_container"] = "MY_CONTAINER"
      LOGGING_CONFIG["handlers"]["task"]["wasb_log_folder"] = "CUSTOM_PREFIX"
```

~~__Finally, you may need to make airflow see `/opt/airflow/config/` as an importable package:__~~

~~If you run into an import error for `airflow_local_settings.LOGGING_CONFIG` you will need to mount an empty `__init__.py` file into the `/opt/airflow/config/` folder. You can do this with the `airflow.extraVolumeMounts` and `airflow.extraVolumes` values.~~

**Reply by HakarK** [reply]
With these values using airflow version 2.6.0:
![image](https://github.com/airflow-helm/charts/assets/60604135/902d9aa5-f77d-477a-adb7-3b5227c06e0b)
![image](https://github.com/airflow-helm/charts/assets/60604135/b849313e-c367-4143-b8ab-96dbd2945957)

login is set to storage account name (airflowstvg) in my_wasb connection.

The pods goes into init:CrashLoopBackOff status.
I have not seen any import error.

Is the value of wasb_log_folder arbitrary? or does it actually need to be the name of the folder inside the container (wasb-airflowlogs)? I made the folder logs in the container, but it still did not work.

**Reply by thesuperzapper** [reply]
@HakarK I think you may have confused my comment, I have updated it to be more clear. Note, for Airflow 2.6, you ONLY need to set the configs, not the `local_settings.py`.

Because I don't have an Azure ready to test, I would greatly appreciate if you can test two things:

1. Confirm that the `"SOLUTION: Airflow 2.6"` listed above works
2. Try using `"SOLUTION: Airflow 2.0 - 2.5"` (in airflow 2.6 is fine), and confirm if it works (and if the `__init__.py` mount is needed)

---

In both solutions, `wasb_log_folder` is the "file prefix" under which the logs will be written.

**Reply by HakarK** [reply]
Solution 1 did not work for me.
Solution 2 works, and I did not have to make the `__init__.py` file. I tested solution 2 with airflow version 2.4.3-python3.8

The documentation should be updated. Thank you for the help!

**Reply by thesuperzapper** [reply]
@HakarK just to clarify, when you tried solution 1 were you using Airflow 2.6? 

Because solution 1 will only work with Airflow 2.6+, but solution 2 should work for all versions (including 2.6).

**Reply by HakarK** [reply]
I tried solution 1 with version 2.6.0 and 2.6.3.

**Comment by HakarK** [comment]
@thesuperzapper  Is there documentation for reading DAGs from storage account?
